### PR TITLE
Match redirect chain URLs against Brave Ad conversion URL patterns

### PR DESCRIPTION
**Pull request by tmancey**
Resolves https://github.com/brave/brave-browser/issues/13033

## Submitter Checklist:

- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed).
- [ ] Requested a security/privacy review as needed.

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

Add a conversion where the landing page has redirects and the url pattern matches an intermediate URL in the redirect chain

### CODE REVIEW

**Review by yachtcaptain23 — APPROVED** [review]

**Comment by tmancey** [comment]
CI failed due to unrelated rewards browser tests

File: vendor/brave-ios/Ads/BATBraveAds.mm [hunk 1]
```diff
@@ -487,7 +487,7 @@ - (void)reportLoadedPageWithURL:(NSURL *)url innerText:(NSString *)text tabId:(N
 {
   if (![self isAdsServiceRunning]) { return; }
   const auto urlString = base::SysNSStringToUTF8(url.absoluteString);
-  ads->OnPageLoaded((int32_t)tabId, urlString, urlString, base::SysNSStringToUTF8(text));
+  ads->OnPageLoaded((int32_t)tabId, { urlString }, base::SysNSStringToUTF8(text));
```
Comment (kylehickinson) [inline vendor/brave-ios/Ads/BATBraveAds.mm:490]: Does iOS need to implement this correctly?

Comment (tmancey) [reply vendor/brave-ios/Ads/BATBraveAds.mm:490]: It does and we were not sure if this was possible so it was left for investigation. I have raised https://github.com/brave/brave-ios/issues/3111

Comment (kylehickinson) [reply vendor/brave-ios/Ads/BATBraveAds.mm:490]: Assuming we don't support it, will there be any specific issues?

Comment (tmancey) [reply vendor/brave-ios/Ads/BATBraveAds.mm:490]: If iOS is unable to support then ad conversions will not work for some advertisers who require the intermediate URLs

**Review by kylehickinson — CHANGES_REQUESTED** [review]
iOS wrapper method `reportLoadedPageWithURL` should be updated to include argument for sending redirect chain so that when it comes time to implement in iOS we do not need to do uplifts

This likely includes updating it in both `BATBraveRewards` and `BATBraveAds`

**Review by kylehickinson — APPROVED** [review]
iOS++

**Comment by btlechowski** [comment]
Verification passed on

Brave | 1.20.60 Chromium: 88.0.4315.7 (Official Build) nightly (64-bit)
-- | --
Revision | 9b020863b2987261f6dcd018c6b388041c035f98-refs/branch-heads/4315@{#15}
OS | Ubuntu 18.04 LTS

Important: Permanent redirects are not supported. Only 307 redirects are supported.

Verified test plan from https://github.com/brave/internal/issues/745
**Convert 307 redirect - no wildcard**
Verified conversion was triggered
```
[12675:12675:0106/161717.262314:VERBOSE1:conversions.cc(230)] Conversion for creative set id 0a697a3b-233a-44bc-a337-e2441fb277a7 and ad_notification
[12675:12675:0106/161717.262969:VERBOSE1:conversions.cc(396)] Convert creative instance id 30ecd70f-f56d-4745-a457-4035fcab54b6 and creative set id 0a697a3b-233a-44bc-a337-e2441fb277a7 in 0 hours, 4 minutes, 48 seconds at 4:22 PM
[12675:12675:0106/161717.287841:VERBOSE6:conversions.cc(290)] Successfully logged conversion event
```
**Convert 307 redirect - wildcard**
Verified conversion was triggered
```
[13509:13509:0106/162808.907162:VERBOSE1:conversions.cc(230)] Conversion for creative set id 0a697a3b-233a-44bc-a337-e2441fb277a7 and ad_notification
[13509:13509:0106/162808.909648:VERBOSE1:conversions.cc(396)] Convert creative instance id 30ecd70f-f56d-4745-a457-4035fcab54b6 and creative set id 0a697a3b-233a-44bc-a337-e2441fb277a7 in 0 hours, 6 minutes, 6 seconds at 4:34 PM
[13509:13509:0106/162808.917387:VERBOSE6:conversions.cc(290)] Successfully logged conversion event
```

**Pull request merged**